### PR TITLE
migrate CI build to github actions

### PR DESCRIPTION
**Pull request by bradh**
As requested on the mailing list, migrates most of the Travis CI build activity to Github Actions.

## Checklist

For all pull requests:

- [X] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [X] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [X] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [N/A] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [N/A] PR for bug fixes and small new features are presented as a single commit
- [N/A] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [X] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [N/A] New unit tests have been added covering the changes
- [X] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [X] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [N/A] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates (screenshots, text)
- [N/A] Commits changing the REST API, or any configuration object, should check if the REST API docs (Swagger YAML files and classic documentation) need to be updated.

### CODE REVIEW

**Comment by aaime** [comment]
This is great progress! The missing bit here seems to be the packaging build (ensuring we can build the war/bin and the extension zip packages without error).

**Comment by bradh** [comment]
> This is great progress! The missing bit here seems to be the packaging build (ensuring we can build the war/bin and the extension zip packages without error).

I think this is done. I've folded the package steps into the oraclejdk11 job (see https://github.com/geoserver/geoserver/pull/4581/checks?check_run_id=1458114425), with a (perhaps optimistic) hope that we can reuse the bits we just built, before removing the -SNAPSHOT jars from the cache.

**Comment by bradh** [comment]
Travis is failing with a ConcurrentModificationException, which looks unrelated. Have kicked that build, but don't think it will be relevant.

**Comment by aaime** [comment]
Oh... do you remember where it was? I've recently changed some code related to concurrency (to get higher scalability) and worried that failure might be related.

**Comment by bradh** [comment]
> Oh... do you remember where it was? I've recently changed some code related to concurrency (to get higher scalability) and worried that failure might be related.

Security core, maybe.

File: .github/workflows/linux.yml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+      with:
+        path: ~/.m2/repository
+        key: gs-${{ runner.os }}-maven-${{ hashFiles('**/pom.xml') }}
+        restore-keys: |
+          gs-${{ runner.os }}-maven-
+    - name: Build with Maven
+      run: mvn -B -U -T3 -fae -Dfmt.skip=true -Dall -T2 -Prelease -f src/pom.xml clean install
+    - name: Build community modules
+      run: mvn -nsu -B -U -T4 -fae -Dfmt.skip=true -DskipTests -Prelease -PcommunityRelease -f src/community/pom.xml clean install
+    - name: Package
```
Comment (aaime) [inline .github/workflows/linux.yml:36]: Ah ha, found it, thanks!

File: .github/workflows/linux.yml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        maven-version: 3.6.3
+    - name: Maven repository caching
+      uses: actions/cache@v2
+      with:
+        path: ~/.m2/repository
+        key: gs-${{ runner.os }}-maven-${{ hashFiles('**/pom.xml') }}
+        restore-keys: |
+          gs-${{ runner.os }}-maven-
+    - name: Build with Maven
+      run: mvn -B -U -T3 -fae -Dfmt.action=check -Dqa -DskipTests=true -Dall -T2 -Prelease -f src/pom.xml clean install
```
Comment (aaime) [inline .github/workflows/linux.yml:67]: Seeing a mix of "-T3" and "-T2" here. Might have been already in the Travis build mind.

File: .travis.yml [hunk 3]
```diff
@@ -49,15 +45,3 @@ matrix:
       dist: trusty
```
Comment (aaime) [inline .travis.yml:45]: Does not need to be trusty in particular. We just need a Java 8 build

Comment (bradh) [reply .travis.yml:45]: Ack. Will move from oraclejdk8 to openjdk8 though.

File: .github/workflows/linux.yml (hunk 2 above)
Comment (bradh) [reply .github/workflows/linux.yml:67]: No, my bad.

**Review by mprins — APPROVED** [review]
:+1:

**Comment by aaime** [comment]
Squash-merged for ease of backport. Thanks for the hard work, much appreciated!

**Comment by bradh** [comment]
Added backports to 2.17.x and 2.18.x

**Pull request merged**